File: python/day10.py
import os
import sys
from typing import List, Set, Tuple


def read_data(filename: str) -> List[str]:
    lines: List[str] = []
    with open(filename) as file:
        lines = [line.strip(" \r\n\t") for line in file.readlines()]
    return lines


def parse_map(lines: List[str]) -> List[List[int]]:
    map: List[List[int]] = []
    for line in lines:
        row: List[int] = [int(char) for char in line]
        map.append(row)
    return map


def traverse(map: List[List[int]], y: int, x: int, destinations: Set[Tuple[int, int]]) -> None:
    assert y >= 0
    assert x >= 0
    if (map[y][x] == 9):
        if isinstance(destinations, set):
            destinations.add((y, x))
        elif isinstance(destinations, list):
            destinations.append((y, x))
        return
    # up
    if y > 0 and map[y][x]+1 == map[y-1][x]:
        traverse(map, y-1, x, destinations)
    # right
    if x < len(map[y])-1 and map[y][x]+1 == map[y][x+1]:
        traverse(map, y, x+1, destinations)
    # down
    if y < len(map)-1 and map[y][x]+1 == map[y+1][x]:
        traverse(map, y+1, x, destinations)
    # right
    if x > 0 and map[y][x]+1 == map[y][x-1]:
        traverse(map, y, x-1, destinations)


def get_score(map: List[List[int]], y: int, x: int) -> int:
    destinations: Set[Tuple[int, int]] = set()
    traverse(map, y, x, destinations)
    return len(destinations)


def part1(map: List[List[int]]) -> int:
    total_score: int = 0
    for y, row in enumerate(map):
        for x, cell in enumerate(row):
            if cell == 0:
                score = get_score(map, y, x)
                total_score += score
    return total_score


def get_rating(map: List[List[int]], y: int, x: int):
    paths = list()
    traverse(map, y, x, paths)
    return len(paths)


def part2(map: List[List[int]]) -> int:
    total_score: int = 0
    for y, row in enumerate(map):
        for x, cell in enumerate(row):
            if cell == 0:
                score = get_rating(map, y, x)
                total_score += score
    return total_score


if __name__ == "__main__":
    input_file: str = "/home/mohitjangra/learning/advent_of_code_2024/inputs/day10.input"
    test_file: str = "/home/mohitjangra/learning/advent_of_code_2024/tests/day10.test"
    sample_file: str = "/home/mohitjangra/learning/advent_of_code_2024/tests/day10.sample"

    mode = os.sys.argv[1] if len(os.sys.argv) > 1 else "test"
    if mode.strip().lower() == "input":
        data: List[str] = read_data(input_file)
        map: List[List[int]] = parse_map(data)

        pt1_result = part1(map)
        assert pt1_result == 646

        pt2_result = part2(map)
        assert pt2_result == 1494
    elif mode.strip().lower() == "test":
        data: List[str] = read_data(test_file)
        map: List[List[int]] = parse_map(data)

        pt1_result = part1(map)
        assert pt1_result == 36

        pt2_result = part2(map)
        assert pt2_result == 81
    elif mode.strip().lower() == "sample":
        data: List[str] = read_data(sample_file)
        map: List[List[int]] = parse_map(data)

        sample_result = part1(map)
        assert sample_result == 1
    else:
        print(f"Usage: {os.sys.argv[0]} [test|input]", file=sys.stderr)

File: zig/day10.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

const Position = struct {
    y: usize,
    x: usize,
};

fn read_data(allocator: Allocator, filename: []const u8) ![]const u8 {
    const file = try std.fs.openFileAbsolute(filename, .{ .mode = .read_only });
    defer file.close();
    const file_stats = try file.stat();
    const buffer: []u8 = try allocator.alloc(u8, file_stats.size);
    const data_read: usize = try file.readAll(buffer);
    std.debug.assert(data_read == file_stats.size);
    return buffer;
}

fn parse_data(allocator: Allocator, data: []const u8) ![][]const u4 {
    var lines_splits = std.mem.split(u8, std.mem.trim(u8, data, " \r\n\t"), "\n");
    var row_count: usize = 0;
    while (lines_splits.next()) |_| : (row_count += 1) {}
    lines_splits.reset();
    const rows = try allocator.alloc([]u4, row_count);
    var row_idx: usize = 0;
    while (lines_splits.next()) |line| : (row_idx += 1) {
        const trimed_line = std.mem.trim(u8, line, " \r\n\t");
        const row = try allocator.alloc(u4, trimed_line.len);
        for (trimed_line, 0..) |char, idx| {
            row[idx] = switch (char) {
                '0' => 0,
                '1' => 1,
                '2' => 2,
                '3' => 3,
                '4' => 4,
                '5' => 5,
                '6' => 6,
                '7' => 7,
                '8' => 8,
                '9' => 9,
                else => return error.InvalidCharacter,
            };
        }
        rows[row_idx] = row;
    }
    return rows;
}

fn traverse_for_score(map: [][]const u4, y: usize, x: usize, destinations: *std.hash_map.HashMap(Position, void, std.hash_map.AutoContext(Position), 80)) !void {
    if (map[y][x] == 9) {
        try destinations.put(Position{ .y = y, .x = x }, {});
        return;
    }
    // up
    if (y > 0 and map[y - 1][x] == map[y][x] + 1) {
        try traverse_for_score(map, y - 1, x, destinations);
    }
    // right
    if (x < map[y].len - 1 and map[y][x + 1] == map[y][x] + 1) {
        try traverse_for_score(map, y, x + 1, destinations);
    }
    // down
    if (y < map.len - 1 and map[y + 1][x] == map[y][x] + 1) {
        try traverse_for_score(map, y + 1, x, destinations);
    }
    // left
    if (x > 0 and map[y][x - 1] == map[y][x] + 1) {
        try traverse_for_score(map, y, x - 1, destinations);
    }
}

fn get_score(allocator: Allocator, map: [][]const u4, y: usize, x: usize) !usize {
    var destinations = std.hash_map.HashMap(Position, void, std.hash_map.AutoContext(Position), 80).init(allocator);
    try traverse_for_score(map, y, x, &destinations);
    return destinations.count();
}

fn part1(allocator: Allocator, map: [][]const u4) !usize {
    var total_score: usize = 0;
    for (map, 0..) |row, y| {
        for (row, 0..) |cell, x| {
            if (cell == 0) {
                total_score += try get_score(allocator, map, y, x);
            }
        }
    }
    return total_score;
}

fn traverse_for_rating(map: [][]const u4, y: usize, x: usize, destinations: *std.ArrayList(Position)) !void {
    if (map[y][x] == 9) {
        try destinations.append(Position{ .y = y, .x = x });
        return;
    }
    // up
    if (y > 0 and map[y - 1][x] == map[y][x] + 1) {
        try traverse_for_rating(map, y - 1, x, destinations);
    }
    // right
    if (x < map[y].len - 1 and map[y][x + 1] == map[y][x] + 1) {
        try traverse_for_rating(map, y, x + 1, destinations);
    }
    // down
    if (y < map.len - 1 and map[y + 1][x] == map[y][x] + 1) {
        try traverse_for_rating(map, y + 1, x, destinations);
    }
    // left
    if (x > 0 and map[y][x - 1] == map[y][x] + 1) {
        try traverse_for_rating(map, y, x - 1, destinations);
    }
}

fn get_rating(allocator: Allocator, map: [][]const u4, y: usize, x: usize) !usize {
    var destinations = std.ArrayList(Position).init(allocator);
    try traverse_for_rating(map, y, x, &destinations);
    return destinations.items.len;
}

fn part2(allocator: Allocator, map: [][]const u4) !usize {
    var total_score: usize = 0;
    for (map, 0..) |row, y| {
        for (row, 0..) |cell, x| {
            if (cell == 0) {
                total_score += try get_rating(allocator, map, y, x);
            }
        }
    }
    return total_score;
}

pub fn main() !void {
    const allocator = std.heap.page_allocator;
    const filename: []const u8 = "/home/mohitjangra/learning/advent_of_code_2024/inputs/day10.input";
    const data = try read_data(allocator, filename);
    defer allocator.free(data);
    const map = try parse_data(allocator, data);
    defer {
        for (map) |row| {
            allocator.free(row);
        }
    }
    const pt1_result = try part1(allocator, map);
    std.debug.assert(pt1_result == 646);
    const pt2_result = try part2(allocator, map);
    std.debug.assert(pt2_result == 1494);
}

test {
    const allocator = std.heap.page_allocator;
    const filename: []const u8 = "/home/mohitjangra/learning/advent_of_code_2024/tests/day10.test";
    const data = try read_data(allocator, filename);
    defer allocator.free(data);
    const map = try parse_data(allocator, data);
    defer {
        for (map) |row| {
            allocator.free(row);
        }
    }
    const pt1_result = try part1(allocator, map);
    try std.testing.expect(pt1_result == 36);

    const pt2_result = try part2(allocator, map);
    try std.testing.expect(pt2_result == 81);
}
